File: concrete/src/main/java/pcb/uwu/amount/derived/termodynamics/Joules.java
package pcb.uwu.amount.derived.termodynamics;

import pcb.uwu.amount.base.Length;
import pcb.uwu.amount.base.Meters;
import pcb.uwu.amount.base.Seconds;
import pcb.uwu.amount.base.Time;
import pcb.uwu.amount.derived.electromagnetism.Coulombs;
import pcb.uwu.amount.derived.electromagnetism.ElectricCharge;
import pcb.uwu.amount.derived.electromagnetism.ElectricPotential;
import pcb.uwu.amount.derived.electromagnetism.Volts;
import pcb.uwu.amount.derived.fundamental.Volume;
import pcb.uwu.amount.derived.mechanics.Force;
import pcb.uwu.amount.derived.mechanics.Newtons;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.derived.termodynamics.EnergyUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.unit.base.MeterUnit.METER;
import static pcb.uwu.unit.base.SecondUnit.SECOND;
import static pcb.uwu.unit.derived.electromagnetism.CoulombUnit.COULOMB;
import static pcb.uwu.unit.derived.electromagnetism.VoltUnit.VOLT;
import static pcb.uwu.unit.derived.fundamental.CubicMeterUnit.CUBIC_METER;
import static pcb.uwu.unit.derived.mechanics.NewtonUnit.NEWTON;
import static pcb.uwu.unit.derived.termodynamics.JouleUnit.JOULE;
import static pcb.uwu.unit.derived.termodynamics.PascalUnit.PASCAL;
import static pcb.uwu.unit.derived.termodynamics.WattUnit.WATT;

public class Joules extends Energy {

	// region constructors

	public Joules(Number value) {
		super(value, JOULE);
	}

	public Joules(Number value, Magnitude magnitude) {
		super(value, magnitude, JOULE);
	}

	public Joules(String value) {
		super(value, JOULE);
	}

	public Joules(String value, Magnitude magnitude) {
		super(value, magnitude, JOULE);
	}

	public Joules(BigDecimal value) {
		super(value, JOULE);
	}

	public Joules(BigDecimal value, Magnitude magnitude) {
		super(value, magnitude, JOULE);
	}

	public Joules(BigDecimalAmount amount) {
		super(amount, JOULE);
	}

	public Joules(BigDecimalAmount amount, Magnitude magnitude) {
		super(amount, magnitude, JOULE);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public Joules plus(UnitAmount<EnergyUnit> other, MathContext mathContext) {
		return new Joules(plusAmount(other, mathContext));
	}

	@Override
	public Joules minus(UnitAmount<EnergyUnit> other, MathContext mathContext) {
		return new Joules(minusAmount(other, mathContext));
	}

	@Override
	public Joules multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Joules(multipliedByScalar(other, mathContext));
	}

	@Override
	public Joules dividedBy(BigDecimal other, MathContext mathContext) {
		return new Joules(dividedByScalar(other, mathContext));
	}

	// endregion

	// region composition

	public Pascals dividedBy(Volume volume, MathContext mathContext) {
		return new Pascals(getAmount().dividedBy(volume.getAmountIn(CUBIC_METER), mathContext));
	}

	public Volume dividedBy(Pressure pressure, MathContext mathContext) {
		return new Volume(getAmount().dividedBy(pressure.getAmountIn(PASCAL), mathContext), CUBIC_METER);
	}

	public Seconds dividedBy(Power power, MathContext mathContext) {
		return new Seconds(getAmount().dividedBy(power.getAmountIn(WATT), mathContext));
	}

	public Watts dividedBy(Time time, MathContext mathContext) {
		return new Watts(getAmount().dividedBy(time.getAmountIn(SECOND), mathContext));
	}

	public Newtons dividedBy(Length length, MathContext mathContext) {
		return new Newtons(getAmount().dividedBy(length.getAmountIn(METER), mathContext));
	}

	public Meters dividedBy(Force force, MathContext mathContext) {
		return new Meters(getAmount().dividedBy(force.getAmountIn(NEWTON), mathContext));
	}

	public Volts dividedBy(ElectricCharge electricCharge, MathContext mathContext) {
		return new Volts(getAmount().dividedBy(electricCharge.getAmountIn(COULOMB), mathContext));
	}

	public Coulombs dividedBy(ElectricPotential electricPotential, MathContext mathContext) {
		return new Coulombs(getAmount().dividedBy(electricPotential.getAmountIn(VOLT), mathContext));
	}

	// endregion
}

File: implementation/src/main/java/pcb/uwu/amount/derived/thermodynamics/Joules.kt
package pcb.uwu.amount.derived.thermodynamics

import pcb.uwu.amount.base.Length
import pcb.uwu.amount.base.Meters
import pcb.uwu.amount.base.Seconds
import pcb.uwu.amount.base.Time
import pcb.uwu.amount.derived.electromagnetism.Coulombs
import pcb.uwu.amount.derived.electromagnetism.ElectricCharge
import pcb.uwu.amount.derived.electromagnetism.ElectricPotential
import pcb.uwu.amount.derived.electromagnetism.Volts
import pcb.uwu.amount.derived.geometry.Volume
import pcb.uwu.amount.derived.mechanics.Force
import pcb.uwu.amount.derived.mechanics.Newtons
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.MeterUnit
import pcb.uwu.unit.base.SecondUnit
import pcb.uwu.unit.derived.electromagnetism.CoulombUnit
import pcb.uwu.unit.derived.electromagnetism.VoltUnit
import pcb.uwu.unit.derived.geometry.CubicMeterUnit
import pcb.uwu.unit.derived.mechanics.NewtonUnit
import pcb.uwu.unit.derived.termodynamics.EnergyUnit
import pcb.uwu.unit.derived.termodynamics.JouleUnit
import pcb.uwu.unit.derived.termodynamics.PascalUnit
import pcb.uwu.unit.derived.termodynamics.WattUnit

class Joules : Energy
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, JouleUnit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, JouleUnit)

    // region UnitAmount

    override fun plus(energy: UnitAmount<EnergyUnit>) =
        Joules(this.amount + (energy to JouleUnit).amount)

    override fun minus(energy: UnitAmount<EnergyUnit>) =
        Joules(this.amount - (energy to JouleUnit).amount)

    override fun times(scalar: Number) =
        Joules(this.amount * scalar)

    override fun div(scalar: Number) =
        Joules(this.amount / scalar)

    // endregion

    // region composition

    override fun div(volume: Volume) =
        Pascals(this.amount / (volume to CubicMeterUnit).amount)

    override fun div(pressure: Pressure) =
        Volume(amount = this.amount / (pressure to PascalUnit).amount,
               unit = CubicMeterUnit)

    operator fun div(power: Power) =
        Seconds(this.amount / (power to WattUnit).amount)

    override fun div(time: Time) =
        Watts(this.amount / (time to SecondUnit).amount)

    override fun div(length: Length) =
        Newtons(this.amount / (length to MeterUnit).amount)

    operator fun div(force: Force) =
        Meters(this.amount / (force to NewtonUnit).amount)

    override fun div(electricCharge: ElectricCharge) =
        Volts(this.amount / (electricCharge to CoulombUnit).amount)

    override fun div(electricPotential: ElectricPotential) =
        Coulombs(this.amount / (electricPotential to VoltUnit).amount)

    // endregion
}
